### DISCUSSION TITLE
Fine-Tuning Tacotron2 LJSpeech on Small Dataset

### DISCUSSION THREAD

**Discussion by staylorVail** [coqui-ai/TTS#2556] [General Q&A]
Hi, I am in the process of trying to fine-tune a Tacotron2 model that I originally trained on the LJSpeech dataset in order to correctly render website URLs. I currently have only a small set of 20 fine-tuning samples to work with, which consist of a female English speaker spelling out various website URLS.I repeated the 20 samples 5 times each to reach the minimum of 100 samples needed for training/fine-tuning.

I modified the original training config to reduce the learning rate from 0.001 to 0.00001, but otherwise left it intact (aside from necessary path changes).

The resulting model, after training for about 2000 steps, fully took on the voice of the new speaker and was (somewhat) able to synthesize components of a URL ("www", "http" "://", ".com", etc.) but fully unable to correctly synthesize any words within a URL that were not in the fine-tuning dataset. In addition, whenever it tried to synthesize general words or phases, such as "hello world", it would only be capable of trying to synthesize the words/sounds: "http" (H-T-T-P), "www" (W-W-W), "://" (colon forward-slash forward-slash), etc. that the original speaker spoke.

In general, it seems like all the features that were learned by the original model were quickly lost by the fine-tuning process. Is my learning rate still too high, do I simply need more (non-repeated) samples, or is there some other parameter to blame? I don't care that much about the resulting model being able to synthesize websites very accurately, but I do aim to have a model that can at least roughly render the components of a URL as separate words, without losing the learned features of the original model.

config.json:
```
{
    "model": "Tacotron2",
    "run_name": "ljspeech-finetune-websites-repeat",
    "run_description": "LJSpeech tacotron2 fine-tuned ONLY on websites.",
    "epochs": 1000,
    "batch_size": 64,
    "eval_batch_size": 16,
    "mixed_precision": false,
    "scheduler_after_epoch": false,
    "run_eval": true,
    "test_delay_epochs": 10,
    "print_eval": false,
    "dashboard_logger": "tensorboard",
    "print_step": 25,
    "plot_step": 100,
    "model_param_stats": false,
    "project_name": null,
    "log_model_step": null,
    "wandb_entity": null,
    "save_step": 10000,
    "checkpoint": true,
    "keep_all_best": false,
    "keep_after": 10000,
    "num_loader_workers": 4,
    "num_eval_loader_workers": 0,
    "use_noise_augment": false,
    "use_language_weighted_sampler": false,
    "output_path": "/opt/modeling/staylor/tts_training/models/tacotron2-ljspeech-websites-repeat/",
    "distributed_backend": "gloo",
    "distributed_url": "tcp://localhost:54321",
    "audio": {
        "fft_size": 1024,
        "win_length": 1024,
        "hop_length": 256,
        "frame_shift_ms": null,
        "frame_length_ms": null,
        "stft_pad_mode": "reflect",
        "sample_rate": 22050,
        "resample": false,
        "preemphasis": 0.0,
        "ref_level_db": 20,
        "do_sound_norm": false,
        "log_func": "np.log10",
        "do_trim_silence": true,
        "trim_db": 60,
        "do_rms_norm": false,
        "db_level": null,
        "power": 1.5,
        "griffin_lim_iters": 60,
        "num_mels": 80,
        "mel_fmin": 50.0,
        "mel_fmax": 7600.0,
        "spec_gain": 1,
        "do_amp_to_db_linear": true,
        "do_amp_to_db_mel": true,
        "signal_norm": true,
        "min_level_db": -100,
        "symmetric_norm": true,
        "max_norm": 4.0,
        "clip_norm": true,
        "stats_path": "/opt/modeling/staylor/tts_training/source/config/tacotron2-LJSpeech-fine-tune-websites-repeat/scale_stats.npy"
    },
    "use_phonemes": true,
    "use_espeak_phonemes": false,
    "phoneme_language": "en-us",
    "compute_input_seq_cache": true,
    "text_cleaner": "phoneme_cleaners",
    "enable_eos_bos_chars": false,
    "test_sentences_file": null,
    "phoneme_cache_path": "/opt/coqui/phoneme_cache/",
    "characters": {
        "pad": "_",
        "eos": "~",
        "bos": "^",
        "characters": "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz!'(),-.:;? ",
        "punctuations": "!'(),-.:;? ",
        "phonemes": "iy\u0268\u0289\u026fu\u026a\u028f\u028ae\u00f8\u0258\u0259\u0275\u0264o\u025b\u0153\u025c\u025e\u028c\u0254\u00e6\u0250a\u0276\u0251\u0252\u1d7b\u0298\u0253\u01c0\u0257\u01c3\u0284\u01c2\u0260\u01c1\u029bpbtd\u0288\u0256c\u025fk\u0261q\u0262\u0294\u0274\u014b\u0272\u0273n\u0271m\u0299r\u0280\u2c71\u027e\u027d\u0278\u03b2fv\u03b8\u00f0sz\u0283\u0292\u0282\u0290\u00e7\u029dx\u0263\u03c7\u0281\u0127\u0295h\u0266\u026c\u026e\u028b\u0279\u027bj\u0270l\u026d\u028e\u029f\u02c8\u02cc\u02d0\u02d1\u028dw\u0265\u029c\u02a2\u02a1\u0255\u0291\u027a\u0267\u02b2\u025a\u02de\u026b"
    },
    "batch_group_size": 4,
    "loss_masking": true,
    "sort_by_audio_len": false,
    "min_seq_len": 6,
    "max_seq_len": 180,
    "compute_f0": false,
    "compute_linear_spec": false,
    "add_blank": false,
    "datasets": [
        {
            "name": "ljspeech",
            "path": "/opt/modeling/tts/datasets/LJSpeech-1.1_plus_finetuning/",
            "meta_file_train": "websites_fine_tune_repeat_metadata.csv",
            "ignored_speakers": null,
            "language": "",
            "meta_file_val": null,
            "meta_file_attn_mask": ""
        }
    ],
    "optimizer": "RAdam",
    "optimizer_params": {
        "betas": [
            0.9,
            0.998
        ],
        "weight_decay": 1e-06
    },
    "lr_scheduler": "NoamLR",
    "lr_scheduler_params": {
        "warmup_steps": 4000
    },
    "test_sentences": [
        "It took me quite a long time to develop a voice, and now that I have it I'm not going to be silent.",
        "Be a voice, not an echo.",
        "I'm sorry Dave. I'm afraid I can't do that.",
        "This cake is great. It's so delicious and moist.",
        "Prior to November 22, 1963.",
        "www.twitter.com",
        "https://www.github.com/coqui-ai/TTS"
    ],
    "use_gst": false,
    "gst": {
        "gst_style_input_wav": null,
        "gst_style_input_weights": null,
        "gst_embedding_dim": 256,
        "gst_use_speaker_embedding": false,
        "gst_num_heads": 4,
        "gst_num_style_tokens": 10
    },
    "gst_style_input": null,
    "num_speakers": 1,
    "num_chars": 130,
    "r": 6,
    "gradual_training": [
        [
            0,
            6,
            256
        ],
        [
            10000,
            4,
            128
        ],
        [
            50000,
            3,
            128
        ],
        [
            100000,
            2,
            128
        ]
    ],
    "memory_size": -1,
    "prenet_type": "original",
    "prenet_dropout": true,
    "prenet_dropout_at_inference": false,
    "stopnet": true,
    "separate_stopnet": true,
    "stopnet_pos_weight": 15.0,
    "max_decoder_steps": 1000,
    "encoder_in_features": 512,
    "decoder_in_features": 512,
    "decoder_output_dim": 80,
    "out_channels": 80,
    "attention_type": "original",
    "attention_heads": null,
    "attention_norm": "sigmoid",
    "attention_win": false,
    "windowing": false,
    "use_forward_attn": false,
    "forward_attn_mask": false,
    "transition_agent": false,
    "location_attn": true,
    "bidirectional_decoder": false,
    "double_decoder_consistency": false,
    "ddc_r": 6,
    "use_speaker_embedding": false,
    "speaker_embedding_dim": 512,
    "use_d_vector_file": false,
    "d_vector_file": false,
    "d_vector_dim": null,
    "lr": 1e-05,
    "grad_clip": 5.0,
    "seq_len_norm": false,
    "decoder_loss_alpha": 0.25,
    "postnet_loss_alpha": 0.25,
    "postnet_diff_spec_alpha": 0.25,
    "decoder_diff_spec_alpha": 0.25,
    "decoder_ssim_alpha": 0.25,
    "postnet_ssim_alpha": 0.25,
    "ga_alpha": 5.0
}
```

Tensorboard graphs from the fine-tuning run:
<img width="1787" alt="Screenshot 2023-04-25 at 2 47 42 PM" src="https://user-images.githubusercontent.com/34282198/234387397-1de10067-e804-4121-bffa-05a54987e783.png">

**Comment by prakharpbuf** [comment]
I would suggest using the entire, or at least a part of the lljspeech dataset along with your own dataset to fine tune the model.

also look into weighted sampling.

it might help

**Reply by staylorVail** [reply]
I did originally try to fine-tune with ~5 hours of the original LJSpeech data combined with the same number (2740) of new (repeated) samples, but the results were comparatively worse than my current attempt. While the features the the original model learned were still present, it was still unable to correctly synthesize a URL, and in some cases the synthesized audio consisted of both the old and new speaker's voices in the same output.

Would it be recommended, if even possible, to fine-tune the single-speaker model into a multispeaker model, or should it be possible to fine-tune my single-speaker model with a new speaker and have the resulting voice still sound like the original speaker, if maybe slightly different?
